replacing loops with iterators, unrolling float loop

khodzha opened the pull request.

khodzha commented on resampler/src/speex.rs:
```diff
@@ -1489,12 +1482,10 @@ fn speex_resampler_process_native(
-    while j < n - 1 {
+    (0..n - 1).for_each(|j| {
         st.mem[mem_idx + j as usize] =
             st.mem[mem_idx + (j as u32 + ilen) as usize];
```
i tried to replace this with enumerating on `st.mem` but failed because its borrowed twice here :man_shrugging:

Luni-4 replied:
Since the second `st.mem` remains the same, we could clone that slice. In this way we don't borrow twice. Something like:
`let mem_clone = &st.mem[mem_idx + n + ilen].clone()`

Luni-4 commented on resampler/src/speex.rs:
```diff
@@ -334,14 +333,14 @@ impl SpeexResamplerState {
+                    .enumerate()
+                    .take(ichunk as usize)
+                    .for_each(|(j, x)| {
+                        *x = in_0[j * (istride_save as usize)] as f32;
```
We can use `.zip(in_0.iter())` here and remove the `j` variable definitely. We can also remove the `enumerate()`

Luni-4 commented on resampler/src/speex.rs:
```diff
@@ -353,18 +352,20 @@ impl SpeexResamplerState {
+            out.iter_mut()
+                .step_by(ostride_save as usize)
+                .enumerate()
+                .for_each(|(j, x)| {
```
We can use `.zip(yselfack.iter())` here and remove the `j` variable definitely

Luni-4 commented on resampler/src/speex.rs:
```diff
@@ -1489,12 +1482,10 @@ fn speex_resampler_process_native(
     let ilen: u32 = *in_len;
-    let mut j: i32 = 0;
-    while j < n - 1 {
+    (0..n - 1).for_each(|j| {
```
Same approach used for the `mem_clone` above

Luni-4 commented on resampler/src/speex.rs:
```diff
@@ -1409,7 +1402,7 @@ fn resampler_basic_direct_double(
         {
             let mut accum: [f64; 4] = [0.0; 4];
-            while j < n {
+            (0..n).step_by(4).for_each(|j| {
```
We can iterate over `sinct` and `iptr` using a `zip()` an then sum the result obtained.

khodzha commented:
not really, bench gives `[10.452 ms 10.509 ms 10.618 ms]` :man_shrugging:

Luni-4 commented:
Looking at the benchmark results, the situation is gotten worse :(

lu-zero commented:
Could you write benchmarks for the single function so we can look what happens on it? (also cargo-asm might shed some light)

khodzha commented:
i added a simpler benchmark which doesnt change rate of output 

i also took a look at C implementation of speexdsp and
- unrolled sse multiplications and additions in `resampler_basic_direct_single` 
- changed vector horizontal sum from hadd to movehl/shuffle

results:
```
resampler_simple_c      time:   [2.0033 ms 2.0225 ms 2.0387 ms]
resampler_simple_rust   time:   [2.7497 ms 2.7794 ms 2.7974 ms] (without unroll and hsum with hadd)
resampler_simple_rust   time:   [2.1840 ms 2.2375 ms 2.3400 ms] (with unroll and hsum with movehl/shuffle)
```

lu-zero commented:
It looks really nice :) there is still some overhead that should go away but it is a fairly good improvement :)

Luni-4 commented:
Great @khodzha! Thanks a lot! :)

khodzha commented:
results for doubles right now:
`resampler_simple_rust_dbl` [7.7673 ms 7.9603 ms 8.1017 ms]
`resampler_simple_c_dbl` [8.0655 ms 8.3503 ms 8.6117 ms]

khodzha commented:
i rebased and squashed commits and marked PR as ready for review if you want to merge it

lu-zero commented:
It seems to still have conflicts. great result :)
